homeassistant/core.py

```diff
@@ -260,6 +260,49 @@ def __str__(self) -> str:
         return self.value
 
 
+class BackgroundTasks:
+    """Class to manage background tasks."""
+
+    def __init__(self) -> None:
+        """Initialize the background task runner."""
+        self._tasks: set[asyncio.Task[Any]] = set()
+        self._loop = asyncio.get_running_loop()
+        self._running = True
+
+    def async_create_task(
+        self,
+        target: Coroutine[Any, Any, _R],
+        name: str,
+    ) -> asyncio.Task[_R]:
+        """Create a task and add it to the set of tasks."""
+        if not self._running:
+            raise RuntimeError("BackgroundTasks is no longer running")
+        task = self._loop.create_task(target, name=name)
+        self._tasks.add(task)
+        task.add_done_callback(self._tasks.remove)
+        return task
+
+    async def async_cancel_all(self) -> None:
+        """Cancel all tasks."""
+        self._running = False
+
+        if not self._tasks:
+            return
+
+        for task in self._tasks:
+            task.cancel()
```

Comment:
Raise an error if the task is done, I guess you need the full task object and check if the task is already canceled/done

Reply:
Tasks remove itself from `self._tasks` if they are done via a done callback.